core/query/sql.ml

```diff
@@ -137,138 +152,162 @@ let order_by_clause n =
-              let tables = List.map (fun (t, x) -> quote t ^ " as " ^ (string_of_table_var x)) tables in
-              let q = "(" ^ sq q ^ ") as " ^ string_of_table_var z in
-              string_of_select fields (q::tables) condition os
-          | _ -> assert false
+  let pr_insert ppf table fields values =
+    let pp_value ppf x = Format.fprintf ppf "(%a)" (Format.pp_print_list ~pp_sep:pp_comma pr_b_ignore_fields) x in
+    Format.fprintf ppf "insert into %a (%a) values %a"
+      pp_constant table
+      (Format.pp_print_list ~pp_sep:pp_comma Format.pp_print_string) fields
+      (Format.pp_print_list ~pp_sep:pp_comma pp_value) values
+  in
+  match q with
+    | UnionAll ([], _) -> Format.fprintf ppf "%a" pp_constant "select 42 as \"@unit@\" where false"
+    | UnionAll ([q], n) -> Format.fprintf ppf "%a%a" pr_q q pp_constant (order_by_clause n)
+    | UnionAll (qs, n) ->
+      let pp_sep_union ppf () = Format.fprintf ppf " union all " in
+      let pp_value ppf x = Format.fprintf ppf "(%a)" pr_q x in
+      Format.fprintf ppf "%a%a"
+        (Format.pp_print_list ~pp_sep:pp_sep_union pp_value) qs
+        pp_constant (order_by_clause n)
+    | Select (fields, [], Constant (Constant.Bool true), _os) ->
+      Format.fprintf ppf "select %a" pr_fields fields
+    | Select (fields, [], condition, _os) ->
+      Format.fprintf ppf "select * from (select %a) as %a where %a"
+        pr_fields fields
+        pp_constant (fresh_dummy_var ())
+        pr_b condition
+    | Select (fields, tables, condition, os) ->
+        (* using quote_field assumes tables contains table names (not nested queries) *)
+        let tables = List.map (fun (t, x) -> Format.asprintf "%a as %s" pp_quote t (string_of_table_var x)) tables in
+        pr_select ppf fields tables condition os
+    | Delete { del_table; del_where } ->
+        pr_delete ppf del_table del_where
+    | Update { upd_table; upd_fields; upd_where } ->
+        pr_update ppf upd_table upd_fields upd_where
+    | Insert { ins_table; ins_fields; ins_records } ->
+        pr_insert ppf ins_table ins_fields ins_records
+    | With (_, q, z, q') ->
+        match q' with
+        | Select (fields, tables, condition, os) ->
+            (* Inline the query *)
+            let tables = List.map (fun (t, x) -> Format.asprintf "%a as %s" pp_quote t (string_of_table_var x)) tables in
+            let pr_q ppf q = pr_query ppf quote ignore_fields q in
+            let q = Format.asprintf "(%s) as %s" (Format.asprintf "%a" pr_q q) (string_of_table_var z) in
+            pr_select ppf fields (q::tables) condition os
+        | _ -> assert false
 
-and string_of_base quote one_table b =
-  let sb = string_of_base quote one_table in
+and pr_base ppf quote one_table b =
+  let string_of_projection quote one_table (var, label) =
+  if one_table then
+    quote label
+  else
+    buffer_concat [string_of_table_var var; "."; (quote label)]
+  in
+  let pr_b ppf b = pr_base ppf quote one_table b in
```

Comment:
The definitions of pr_b and pr_q here shadow earlier ones, which might be confusing.